### PR TITLE
Add an enumerate() method

### PR DESCRIPTION
**Pull request by nakulj**

### CODE REVIEW

**Review by fluentfuture — COMMENTED** [review]
Have you considered MoreStreams.indexesFrom()?

Then you can:

```
BiStream.zip(indexesFrom(0), collection.stream());
```

It's slightly more verbose. But also more explicit about:

1. Which of the two is the index (particularly if the stream itself is `Stream<Integer>`)
2. Is it 0-based or 1-based.

**Comment by nakulj** [comment]
I did consider it, but `Stream`'s  `count`, `limit`, and `skip` methods use `long`, while `indexesFrom` is `int`. Someone looking at the `Stream` api alone might reasonably try to `enumerate` a collection with >= 2^32 elements and be surprised by overflow.

Perhaps `indexesFrom` should be rewritten to use `long`?

**Comment by fluentfuture** [comment]
We do internally have an `indexesFrom(long)` overload. Just added it to MoreStreams.

Xo you can do: `zip(indexesFrom(0L), ...)`. And you have full control over int vs. long, 0-based vs. 1-based, `(index, value)` vs. `(value, index)`.

**Comment by nakulj** [comment]
Ok, done!

**Comment by fluentfuture** [comment]
Sorry. I meant, could the new `indexesFrom(long)` satisfy your use case without needing a new method?

**Comment by nakulj** [comment]
Oh I see, sorry for the confusion :)

Yes, this was intended as a convenience method. As you said, it is a little less explicit, but iterating over `index,element` pairs beginning at zero is a very common pattern in many languages:

- Golang's [`range`](https://gobyexample.com/range)
- Rust's [`enumerate`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.enumerate)
- Python's [`enumerate`](https://docs.python.org/3/library/functions.html#enumerate) (interestingly this one does have an optional step parameter)

So I don't think the order of the pairing is a real concern here.

**Comment by fluentfuture** [comment]
We've had similar discussions internally and haven't found a convincing reason to add `enumerate()1.

Guava's `mapWithIndex()` expects a lambda with the second parameter as the index, for example.

There doesn't appear to be a universal preceden in terms of ordering.

Internally, we've also seen people needing 1-based.

Lastly, sometimes using `long` for the index can be cumbersome because the user needs to pass it to another API that accepts int, and they'd have to do a cast.

The bottom line is, `zip(indexesFrom(0L), stream)` isn't less readable, or _that_ inconvenient.

Will close for now.
